api/python/tests/integration/test_workflows.py

```diff
@@ -0,0 +1,345 @@
+                with pytest.raises(QuiltException, match=r'Workflow is required, but none specified.'):
+                    self._validate(workflow=workflow)
+
+    def test_workflow_is_required_default_set(self):
+        set_local_conf_data(get_v1_conf_data('''
+            default_workflow: w1
+            workflows:
+              w1:
+                name: Name
+        '''))
+
+        assert self._validate() == {
+            'id': 'w1',
+            'config': str(get_package_registry().workflow_conf_pk),
+        }
+
+        with pytest.raises(QuiltException, match=r'Workflow is required, but none specified.'):
+            self._validate(workflow=None)
+
+    def test_workflow_not_required_not_specified(self):
+        set_local_conf_data(get_v1_conf_data('''
+            is_workflow_required: false
+            workflows:
+              w1:
+                name: Name
+        '''))
+        for workflow in (None, ...):
+            with self.subTest(workflow=workflow):
+                assert self._validate(workflow=workflow) == {
+                    'id': None,
+                    'config': str(get_package_registry().workflow_conf_pk),
+                }
+
+    def test_workflow_not_required_default_set(self):
+        set_local_conf_data(get_v1_conf_data('''
+            is_workflow_required: false
+            default_workflow: w1
+            workflows:
+              w1:
+                name: Name
+        '''))
+
+        assert self._validate() == {
+            'id': 'w1',
+            'config': str(get_package_registry().workflow_conf_pk),
+        }
+
+        assert self._validate(workflow=None) == {
+            'id': None,
+            'config': str(get_package_registry().workflow_conf_pk),
+        }
+
+    def test_missing_workflow(self):
+        set_local_conf_data(get_v1_conf_data('''
+            workflows:
+              w1:
+                name: Name
+        '''))
+
+        with pytest.raises(QuiltException, match=r"There is no 'w2' workflow in config."):
+            self._validate(workflow='w2')
+
+    def test_missing_schema(self):
+        set_local_conf_data(get_v1_conf_data('''
+            workflows:
+              w1:
+                name: Name
+                metadata_schema: schema-id
+        '''))
+
+        with pytest.raises(QuiltException, match=r"There is no 'schema-id' in schemas."):
+            self._validate(workflow='w1')
+
+    def test_schema_with_ref(self):
+        set_local_conf_data(get_v1_conf_data('''
+            workflows:
+              w1:
+                name: Name
+                metadata_schema: schema-id
+            schemas:
+              schema-id:
+                url: %s
+        ''' % create_local_tmp_schema('{"$ref": "other-schema"}')))
+
+        with pytest.raises(QuiltException, match=r"Currently we don't support \$ref in schema."):
+            self._validate(workflow='w1')
+
+    def test_schema_validation_invalid_meta(self):
```

Comment:
This feels like one of the most important tests; shouldn't there be at least on test against `push` that hits this case?

Reply:
I added more unit tests in test_packages.py.